### PR TITLE
Clean up history hierarchy

### PR DESCRIPTION
**Pull request by yannikschaelte**
* Properly mark abstract base class methods as abstract instead of NotImplementedError, to not get surprises about missing methods at runtime.
* Refactor history hierarchy, as previously e.g. hdf5history was re-implementing and not using existing fields.

### CODE REVIEW

File: pypesto/history/base.py [hunk 1]
```diff
@@ -246,51 +251,117 @@ def get_trimmed_indices(self) -> np.ndarray:
...
-    Parameters
-    ----------
-    options:
-        History options.
+    def update(  # noqa: D102
+        self,
+        x: np.ndarray,
+        sensi_orders: Tuple[int, ...],
+        mode: ModeType,
+        result: ResultDict,
```
Comment (dweindl) [inline pypesto/history/base.py:276]:
```suggestion
        *args, 
        **kwargs,
```
Also for all other not-planned-to-ever-be-implemented methods?

File: pypesto/history/base.py [hunk 2]
```diff
@@ -246,51 +251,117 @@ def get_trimmed_indices(self) -> np.ndarray:
...
+    Can be used whenever a history object is needed, but no history is desired.
+    Can be created, but not queried.
     """
-    Tracks number of function evaluations only, no trace.
 
-    Parameters
-    ----------
-    options:
-        History options.
+    def update(  # noqa: D102
```
Comment (dweindl) [inline pypesto/history/base.py:271]: Those `# noqa: D102` are so ugly. Is there no way around that to not enforce docstrings in overriding methods?

**Review by dweindl — APPROVED** [review]

File: pypesto/history/base.py (hunk 1 above)
Comment (yannikschaelte) [reply pypesto/history/base.py:276]: I think it makes sense to define the function header in as much detail as possible, right?

File: pypesto/history/base.py (hunk 2 above)
Comment (yannikschaelte) [reply pypesto/history/base.py:271]: Yes, I agree. I played around with https://pypi.org/project/overrides/ and `flake8 --ignore-decorators=overrides` for a moment, but got recursion errors and decided it's not worth it ...

Comment (yannikschaelte) [reply pypesto/history/base.py:271]: or now that most functions are documented, we could ignore D102 again, and just pay attention with new prs ... would have no problem with that per se ...

Comment (dweindl) [reply pypesto/history/base.py:271]: Maybe safer with D102... Okay to leave.

File: pypesto/history/base.py (hunk 1 above)
Comment (dweindl) [reply pypesto/history/base.py:276]: Fine to keep, but doesn't feel helpful here. args/kwargs is less verbose and won't require changes if the base signature changes.

**Pull request merged**